sound/soc/sof/topology.c

```diff
@@ -1805,21 +1805,27 @@ static int sof_widget_load_mux(struct snd_soc_component *scomp, int index,
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_soc_tplg_private *private = &tw->priv;
 	struct sof_ipc_comp_mux *mux;
+	size_t ipc_size = sizeof(*mux) + sizeof(swidget->comp_ext);
 	int ret;
 
-	mux = kzalloc(sizeof(*mux), GFP_KERNEL);
+	mux = kzalloc(ipc_size, GFP_KERNEL);
 	if (!mux)
 		return -ENOMEM;
 
 	/* configure mux IPC message */
-	mux->comp.hdr.size = sizeof(*mux);
+	mux->comp.hdr.size = ipc_size;
 	mux->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
 	mux->comp.id = swidget->comp_id;
 	mux->comp.type = SOF_COMP_MUX;
 	mux->comp.pipeline_id = index;
 	mux->comp.core = core;
 	mux->config.hdr.size = sizeof(mux->config);
 
+	/* append extended data to the end of the component */
+	memcpy((u8 *)mux + sizeof(*mux),
+	       &swidget->comp_ext, sizeof(swidget->comp_ext));
+	mux->comp.ext_data_length = sizeof(swidget->comp_ext);
```

Comment:
could we have a helper for all this? It's the same thing done in multiple places.

Reply:
I thought about doing this in sof_widget_ready() with common code and pass it to each specific widget_load, but has no chance to implement it yet. @marc-hb could you please help to follow up this?

Reply:
@plbossart @marc-hb I just added helper for this, and add the handle of no extended data cases there also, please help review.